src/bidiMapper/domains/context/bidiPreloadScript.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2023 Google LLC.
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import type {CommonDataTypes, Script} from '../../../protocol/protocol.js';
+import {uuidv4} from '../../../utils/uuid.js';
+import type {CdpTarget} from './cdpTarget.js';
+
+type CdpPreloadScript = {
+  /** CDP target. Includes session ID and target ID. */
+  target: CdpTarget;
+  /** CDP preload script ID. */
+  preloadScriptId: Script.PreloadScript;
+};
+
+/**
+ * BiDi IDs are generated by the server and are unique within the context.
+ *
+ * CDP preload script IDs are generated by the client and are unique
+ * within the session.
+ *
+ * The mapping between BiDi and CDP preload script IDs is 1:many.
+ * BiDi IDs are needed by the mapper to keep track of potential multiple CDP IDs
+ * in the client.
+ */
+export class BidiPreloadScript {
+  /** BiDi ID, an automatically generated UUID. */
+  readonly #id: string = uuidv4();
+  /** CDP preload scripts. */
+  #cdpPreloadScripts: CdpPreloadScript[] = [];
+  /** The script itself, in a format expected by the spec i.e. a function. */
+  readonly #functionDeclaration: string;
+  /** Browsing context ID. */
+  readonly #contextId: CommonDataTypes.BrowsingContext | null;
+  readonly #targetIds: Set<string> = new Set();
+
+  get id(): string {
+    return this.#id;
+  }
+
+  get contextId(): CommonDataTypes.BrowsingContext | null {
+    return this.#contextId;
+  }
+
+  get targetIds(): Set<string> {
+    return this.#targetIds;
+  }
+
+  constructor(params: Script.AddPreloadScriptParameters) {
+    if (params.sandbox !== undefined) {
+      // TODO: Handle sandbox.
+      throw new Error('Sandbox is not supported yet');
+    }
+    if (params.arguments !== undefined && params.arguments.length > 0) {
+      // TODO: Handle arguments.
+      throw new Error('add preload script arguments are not supported');
+    }
+
+    this.#functionDeclaration = params.functionDeclaration;
+    this.#contextId = params.context ?? null;
+  }
+
+  async initInTargets(cdpTargets: Set<CdpTarget>) {
```

Comment:
I'd argue this can be less strict than `Set`.

`Iterable` is more forgiving. Or just a plain array.

It's totally fine to create a `Set` within this function for the sake of deduplication, but let's make the interface more forgiving

(i.e. basically https://en.wikipedia.org/wiki/Robustness_principle)

Reply:
done